src/main/java/com/eerussianguy/blazemap/feature/maps/MapRenderer.java

```diff
@@ -432,17 +436,40 @@ private void generateMapTileAsync(NativeImage texture, TileResolution resolution
     private void generateMapTile(NativeImage texture, TileResolution resolution, int textureW, int textureH, int cornerXOffset, int cornerZOffset, int regionIndexX, int regionIndexZ) {
         for(BlazeRegistry.Key<Layer> layer : visible) {
             if(layer.value() instanceof FakeLayer) return;
+
+            // Precomputing values so they don't waste CPU cycles recalculating for each pixel
             final RegionPos region = offsets[regionIndexX][regionIndexZ];
             final int cxo = cornerXOffset / resolution.pixelWidth;
             final int czo = cornerZOffset / resolution.pixelWidth;
+
+            final int regWidthbyIndexX = (regionIndexX * resolution.regionWidth);
+            final int regWidthbyIndexZ = (regionIndexZ * resolution.regionWidth);
+
             tileStorage.consumeTile(layer, region, resolution, source -> {
-                for(int x = (region.x * 512) < begin.getX() ? cxo : 0; x < source.getWidth(); x++) {
-                    int textureX = (regionIndexX * resolution.regionWidth) + x - cxo;
-                    if(textureX < 0 || textureX >= textureW) continue;
+                final int sourceWidth = source.getWidth();
+                final int sourceHeight = source.getHeight();
+
+                int startX = (region.x * 512) < begin.getX() ? cxo : 0;
+                int startY = (region.z * 512) < begin.getZ() ? czo : 0;
+
+                if (regWidthbyIndexX + startX - cxo < 0) {
+                    // Set x to be the value it should be when textureX == 0
+                    startX = czo - regWidthbyIndexX;
+                }
+                if (regWidthbyIndexZ + startY - czo < 0) {
+                    // Set y to be the value it should be when textureY == 0
+                    startY = czo - regWidthbyIndexZ;
+                }
+
+                for(int x = startX; x < sourceWidth; x++) {
+                    int textureX = regWidthbyIndexX + x - cxo;
+
+                    if(textureX >= textureW) break;
```

Comment:
Now we don't have to worry about the `textureX < 0` case, we can just `break` when we reach the end of the area.